taichi/codegen/codegen_opengl.cpp

```diff
@@ -166,98 +167,191 @@ class KernelGen : public IRVisitor {
-          "#define _arg_f64(x) _args_f64_[(x) << 0]\n";
+          "layout(std430, binding = 1) buffer args_f32 { float _args_f32_[]; };\n"
+          "layout(std430, binding = 1) buffer args_f64 { double _args_f64_[]; };\n";
     }
     if (used.global_temp) {
       kernel_header +=
           "layout(std430, binding = 2) buffer gtmp_i32 { int _gtmp_i32_[]; };\n"
-          "layout(std430, binding = 2) buffer gtmp_f32 { float _gtmp_f32_[]; "
-          "};\n"
-          "layout(std430, binding = 2) buffer gtmp_f64 { double _gtmp_f64_[]; "
-          "};\n"
-          "#define _gtx_i32(x) _gtmp_i32_[(x) >> 2]\n"
-          "#define _gtx_f32(x) _gtmp_f32_[(x) >> 2]\n"
-          "#define _gtx_f64(x) _gtmp_f64_[(x) >> 3]\n";
+          "layout(std430, binding = 2) buffer gtmp_f32 { float _gtmp_f32_[]; };\n"
+          "layout(std430, binding = 2) buffer gtmp_f64 { double _gtmp_f64_[]; };\n";
     }
     if (used.extra_arg) {
       kernel_header +=
-          "layout(std430, binding = 3) buffer earg_i32 { int _earg_i32_[]; "
-          "};\n" +
-          fmt::format("#define _extra_arg(i, j) _earg_i32_[(i) * {} + (j)]\n",
-                      taichi_max_num_indices);
+          "layout(std430, binding = 3) buffer earg_i32 { int _earg_i32_[]; };\n";
     }
     if (used.external_ptr) {
       kernel_header +=
           "layout(std430, binding = 4) buffer extr_i32 { int _extr_i32_[]; };\n"
-          "layout(std430, binding = 4) buffer extr_f32 { float _extr_f32_[]; "
-          "};\n"
-          "layout(std430, binding = 4) buffer extr_f64 { double _extr_f64_[]; "
-          "};\n"
-          "#define _ext_ns_i32(x) _extr_i32_[(x) >> 0]\n"
-          "#define _ext_ns_f32(x) _extr_f32_[(x) >> 0]\n"
-          "#define _ext_ns_f64(x) _extr_f64_[(x) >> 0]\n";
+          "layout(std430, binding = 4) buffer extr_f32 { float _extr_f32_[]; };\n"
+          "layout(std430, binding = 4) buffer extr_f64 { double _extr_f64_[]; };\n";
     }
-    kernel_header +=
-        "#define _Ax_(x) x\n"
-        "#define _At_(x) _Ax_(_at_##x(x))\n";
-    if (used.atomic_float) {  // {{{
+    if (used.atomic_float && !opengl_has_GL_NV_shader_atomic_float) {  // {{{
       kernel_header +=
           "\
-#define _Atmf_Def(Add, _f_, _o_, mem, _32, float) \
-float atomic##Add##_##mem##_f##_32(int addr, float rhs) \
+float atomicAdd_data_f32(int addr, float rhs) \
+{ \
+  int old, new, ret; \
+  do { \
+    old = _data_i32_[addr]; \
+    new = floatBitsToInt((intBitsToFloat(old) + rhs)); \
+  } while (old != atomicCompSwap(_data_i32_[addr], old, new)); \
+  return intBitsToFloat(old); \
+}\n\
+float atomicSub_data_f32(int addr, float rhs) \
+{ \
+  int old, new, ret; \
+  do { \
+    old = _data_i32_[addr]; \
+    new = floatBitsToInt((intBitsToFloat(old) - rhs)); \
+  } while (old != atomicCompSwap(_data_i32_[addr], old, new)); \
+  return intBitsToFloat(old); \
+}\n\
+float atomicMax_data_f32(int addr, float rhs) \
```

Comment:
Would it be possible to use the `STR` macro by @k-ye so that you get rid of `\` and `\n`? https://github.com/taichi-dev/taichi/blob/62cdc60f2e3d981b4d5c1c9f60b7590c9e8031a1/taichi/backends/metal/shaders/runtime_kernels.metal.h#L44-L70

Reply:
Cool!